## SF crime data analysis

**In this notebook, I use Spark SQL for big data analysis on SF crime data from 2003 to 2018/05/15.** 

(https://data.sfgov.org/Public-Safety/Police-Department-Incident-Reports-Historical-2003/tmnf-yvry). 

The current version is OLAP (Online Analytical Processing) for SF crime data analysis.  

(--Possible plans in the future:
1. unsupervised learning for spatial data analysis.  
2. time series data analysis.  )

In [0]:
from csv import reader
from pyspark.sql import Row 
from pyspark.sql import SparkSession
from pyspark.sql.types import *
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt
import warnings

import os
os.environ["PYSPARK_PYTHON"] = "python3"


In [0]:
# download data from SF gov website

#import urllib.request
#urllib.request.urlretrieve("https://data.sfgov.org/api/views/tmnf-yvry/rows.csv?accessType=DOWNLOAD", "/tmp/myxxxx.csv")
#dbutils.fs.mv("file:/tmp/myxxxx.csv", "dbfs:/laioffer/spark_hw1/data/sf_03_18.csv")
#display(dbutils.fs.ls("dbfs:/laioffer/spark_hw1/data/"))
## link
# https://data.sfgov.org/api/views/tmnf-yvry/rows.csv?accessType=DOWNLOAD


In [0]:
data_path = "dbfs:/laioffer/spark_hw1/data/sf_03_18.csv"
# use this file name later

In [0]:

from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("crime analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

sf_crime = spark.read.format("csv")\
  .option("header", "true")\
  .option("interSchema", "true")\
  .load(data_path)

#original dataset
display(sf_crime)


PdId IncidntNum Incident Code Category Descript DayOfWeek Date Time PdDistrict Resolution Address X Y location SF Find Neighborhoods 2 2 Current Police Districts 2 2 Current Supervisor Districts 2 2 Analysis Neighborhoods 2 2 DELETE - Fire Prevention Districts 2 2 DELETE - Police Districts 2 2 DELETE - Supervisor Districts 2 2 DELETE - Zip Codes 2 2 DELETE - Neighborhoods 2 2 DELETE - 2017 Fix It Zones 2 2 Civic Center Harm Reduction Project Boundary 2 2 Fix It Zones as of 2017-11-06 2 2 DELETE - HSOC Zones 2 2 Fix It Zones as of 2018-02-07 2 2 CBD, BID and GBD Boundaries as of 2017 2 2 Areas of Vulnerability, 2016 2 2 Central Market/Tenderloin Boundary 2 2 Central Market/Tenderloin Boundary Polygon - Updated 2 2 HSOC Zones as of 2018-06-05 2 2 OWED Public Spaces 2 2 Neighborhoods 2 3114751606302 031147516 06302 LARCENY/THEFT PETTY THEFT FROM A BUILDING Sunday 09/28/2003 10:00 SOUTHERN NONE 1ST ST / BRANNAN ST -120.5 90 POINT (-120.50000000000001 90) null null null null null null null null null null null null null null null null null null null null null 5069701104134 050697011 04134 ASSAULT BATTERY Wednesday 06/22/2005 12:20 NORTHERN NONE 1200 Block of EDDY ST -122.428223303176 37.7818959488603 POINT (-122.42822330317601 37.7818959488603) 97 4 11 39 7 9 11 29490 41 null null null null null null 2 null null null null 97 6074729204104 060747292 04104 ASSAULT ASSAULT Saturday 07/15/2006 00:55 CENTRAL NONE 1600 Block of POWELL ST -122.410672425337 37.799788690123 POINT (-122.41067242533701 37.799788690123) 106 6 3 6 3 1 10 308 4 null null null null null null 2 null null null null 106 7103536315201 071035363 15201 ASSAULT STALKING Tuesday 09/25/2007 00:01 TARAVAL NONE 400 Block of ULLOA ST -122.458226300605 37.7413616001449 POINT (-122.458226300605 37.7413616001449) 49 10 8 41 1 8 4 59 40 null null null null null null 1 null null null null 49 11082415274000 110824152 74000 MISSING PERSON MISSING ADULT Saturday 09/24/2011 11:00 TARAVAL LOCATED 3200 Block of SAN JOSE AV -122.459172646607 37.7082001648459 POINT (-122.459172646607 37.7082001648459) null null null null null null null 28588 null null null null null null null null null null null null null 4037801104134 040378011 04134 ASSAULT BATTERY Friday 12/12/2003 12:00 SOUTHERN NONE 500 Block of I-80 -122.386667033903 37.7898821569191 POINT (-122.38666703390301 37.7898821569191) null null null null null null null 28856 null null null null null null null null null null null null null 4147669007025 041476690 07025 VEHICLE THEFT STOLEN TRUCK Thursday 12/30/2004 19:00 BAYVIEW NONE 100 Block of KAREN CT -122.407531316146 37.723291086507 POINT (-122.40753131614602 37.723291086507) 91 2 2 25 10 3 7 309 28 null null null null null null 1 null null null null 91 16010127305073 160101273 05073 BURGLARY BURGLARY, UNLAWFUL ENTRY Wednesday 02/03/2016 20:30 MISSION ARREST, BOOKED 2300 Block of 16TH ST -122.40952991327822 37.76571843952822 POINT (-122.40952991327822 37.76571843952822) 53 3 9 20 8 4 8 28853 19 null null null 3 null null 2 null null 3 null 53 17004924306243 170049243 06243 LARCENY/THEFT PETTY THEFT FROM LOCKED AUTO Wednesday 01/18/2017 18:20 TARAVAL NONE NORIEGA ST / SUNSET BL -122.49522504431104 37.75341185338174 POINT (-122.49522504431104 37.75341185338174) 39 10 7 35 1 8 3 56 35 null null null null null null 1 null null null null 39 16065828006244 160658280 06244 LARCENY/THEFT GRAND THEFT FROM LOCKED AUTO Sunday 08/14/2016 21:00 TARAVAL NONE 3200 Block of 20TH AV -122.47564746078615 37.72852806274646 POINT (-122.47564746078615 37.72852806274647) 41 10 8 16 1 8 4 64 14 null null null null null null 2 null null null null 41 18032260316100 180322603 16100 DRUG/NARCOTIC POSSESSION OF HEROIN Tuesday 05/01/2018 13:27 SOUTHERN ARREST, BOOKED 600 Block of STEVENSON ST -122.41265189435406 37.77951316688882 POINT (-122.41265189435406 37.77951316688882) 32 5 10 34 8 2 9 28853 34 3 1 3 1 3 7 2 1 1 1 null 32 17612518006244 176125180 06244 LARCENY/THEFT GRAND THEFT FROM LOCKED AUTO Tuesday 05/16

In [0]:
## function to transform the date
from pyspark.sql.functions import to_date, to_timestamp, hour, month, year
sf_crime = sf_crime.withColumn('IncidntDate', to_date(sf_crime.Date, "MM/dd/yyyy"))
sf_crime = sf_crime.withColumn('Time', to_timestamp(sf_crime.Time, "HH:mm"))
sf_crime = sf_crime.withColumn('Hour', hour(sf_crime['Time']))
#df_opt1 = sf_crime.withColumn("DayOfWeek", date_format(sf_crime.Date, "EEEE"))
sf_crime = sf_crime.withColumn('Year', year(sf_crime['IncidntDate']))
sf_crime = sf_crime.withColumn('Month', month(sf_crime['IncidntDate']))

# dataset after time transformation
# create sql table
sf_crime.createOrReplaceTempView("sf_crime")
display(sf_crime)
sf_crime.cache()

PdId IncidntNum Incident Code Category Descript DayOfWeek Date Time PdDistrict Resolution Address X Y location SF Find Neighborhoods 2 2 Current Police Districts 2 2 Current Supervisor Districts 2 2 Analysis Neighborhoods 2 2 DELETE - Fire Prevention Districts 2 2 DELETE - Police Districts 2 2 DELETE - Supervisor Districts 2 2 DELETE - Zip Codes 2 2 DELETE - Neighborhoods 2 2 DELETE - 2017 Fix It Zones 2 2 Civic Center Harm Reduction Project Boundary 2 2 Fix It Zones as of 2017-11-06 2 2 DELETE - HSOC Zones 2 2 Fix It Zones as of 2018-02-07 2 2 CBD, BID and GBD Boundaries as of 2017 2 2 Areas of Vulnerability, 2016 2 2 Central Market/Tenderloin Boundary 2 2 Central Market/Tenderloin Boundary Polygon - Updated 2 2 HSOC Zones as of 2018-06-05 2 2 OWED Public Spaces 2 2 Neighborhoods 2 IncidntDate Hour Year Month 3114751606302 031147516 06302 LARCENY/THEFT PETTY THEFT FROM A BUILDING Sunday 09/28/2003 1970-01-01T10:00:00.000+0000 SOUTHERN NONE 1ST ST / BRANNAN ST -120.5 90 POINT (-120.50000000000001 90) null null null null null null null null null null null null null null null null null null null null null 2003-09-28 10 2003 9 5069701104134 050697011 04134 ASSAULT BATTERY Wednesday 06/22/2005 1970-01-01T12:20:00.000+0000 NORTHERN NONE 1200 Block of EDDY ST -122.428223303176 37.7818959488603 POINT (-122.42822330317601 37.7818959488603) 97 4 11 39 7 9 11 29490 41 null null null null null null 2 null null null null 97 2005-06-22 12 2005 6 6074729204104 060747292 04104 ASSAULT ASSAULT Saturday 07/15/2006 1970-01-01T00:55:00.000+0000 CENTRAL NONE 1600 Block of POWELL ST -122.410672425337 37.799788690123 POINT (-122.41067242533701 37.799788690123) 106 6 3 6 3 1 10 308 4 null null null null null null 2 null null null null 106 2006-07-15 0 2006 7 7103536315201 071035363 15201 ASSAULT STALKING Tuesday 09/25/2007 1970-01-01T00:01:00.000+0000 TARAVAL NONE 400 Block of ULLOA ST -122.458226300605 37.7413616001449 POINT (-122.458226300605 37.7413616001449) 49 10 8 41 1 8 4 59 40 null null null null null null 1 null null null null 49 2007-09-25 0 2007 9 11082415274000 110824152 74000 MISSING PERSON MISSING ADULT Saturday 09/24/2011 1970-01-01T11:00:00.000+0000 TARAVAL LOCATED 3200 Block of SAN JOSE AV -122.459172646607 37.7082001648459 POINT (-122.459172646607 37.7082001648459) null null null null null null null 28588 null null null null null null null null null null null null null 2011-09-24 11 2011 9 4037801104134 040378011 04134 ASSAULT BATTERY Friday 12/12/2003 1970-01-01T12:00:00.000+0000 SOUTHERN NONE 500 Block of I-80 -122.386667033903 37.7898821569191 POINT (-122.38666703390301 37.7898821569191) null null null null null null null 28856 null null null null null null null null null null null null null 2003-12-12 12 2003 12 4147669007025 041476690 07025 VEHICLE THEFT STOLEN TRUCK Thursday 12/30/2004 1970-01-01T19:00:00.000+0000 BAYVIEW NONE 100 Block of KAREN CT -122.407531316146 37.723291086507 POINT (-122.40753131614602 37.723291086507) 91 2 2 25 10 3 7 309 28 null null null null null null 1 null null null null 91 2004-12-30 19 2004 12 16010127305073 160101273 05073 BURGLARY BURGLARY, UNLAWFUL ENTRY Wednesday 02/03/2016 1970-01-01T20:30:00.000+0000 MISSION ARREST, BOOKED 2300 Block of 16TH ST -122.40952991327822 37.76571843952822 POINT (-122.40952991327822 37.76571843952822) 53 3 9 20 8 4 8 28853 19 null null null 3 null null 2 null null 3 null 53 2016-02-03 20 2016 2 17004924306243 170049243 06243 LARCENY/THEFT PETTY THEFT FROM LOCKED AUTO Wednesday 01/18/2017 1970-01-01T18:20:00.000+0000 TARAVAL NONE NORIEGA ST / SUNSET BL -122.49522504431104 37.75341185338174 POINT (-122.49522504431104 37.75341185338174) 39 10 7 35 1 8 3 56 35 null null null null null null 1 null null null null 39 2017-01-18 18 2017 1 16065828006244 160658280 06244 LARCENY/THEFT GRAND THEFT FROM LOCKED AUTO Sunday 08/14/2016 1970-01-01T21:00:00.000+0000 TARAVAL NONE 3200 Block of 20TH AV -122.47564746078615 37.72852806274646 POINT (-122.47564746078615 37.7285280627464

Out[5]: DataFrame[PdId: string, IncidntNum: string, Incident Code: string, Category: string, Descript: string, DayOfWeek: string, Date: string, Time: timestamp, PdDistrict: string, Resolution: string, Address: string, X: string, Y: string, location: string, SF Find Neighborhoods 2 2: string, Current Police Districts 2 2: string, Current Supervisor Districts 2 2: string, Analysis Neighborhoods 2 2: string, DELETE - Fire Prevention Districts 2 2: string, DELETE - Police Districts 2 2: string, DELETE - Supervisor Districts 2 2: string, DELETE - Zip Codes 2 2: string, DELETE - Neighborhoods 2 2: string, DELETE - 2017 Fix It Zones 2 2: string, Civic Center Harm Reduction Project Boundary 2 2: string, Fix It Zones as of 2017-11-06 2 2: string, DELETE - HSOC Zones 2 2: string, Fix It Zones as of 2018-02-07 2 2: string, CBD, BID and GBD Boundaries as of 2017 2 2: string, Areas of Vulnerability, 2016 2 2: string, Central Market/Tenderloin Boundary 2 2: string, Central Market/Tenderloin Boundary Polygon - Updated 2 2: string, HSOC Zones as of 2018-06-05 2 2: string, OWED Public Spaces 2 2: string, Neighborhoods 2: string, IncidntDate: date, Hour: int, Year: int, Month: int]

####0. number of crimes by year

In [0]:
crime_yearly = spark.sql("select count(*) as yearly_crimes, Year from sf_crime group by 2 order by 2")
display(crime_yearly)

yearly_crimes Year 142803 2003 142054 2004 137048 2005 131856 2006 131771 2007 135242 2008 134309 2009 127758 2010 126713 2011 135464 2012 147664 2013 144844 2014 151459 2015 145994 2016 149487 2017 45059 2018

#####Analysis
The line chart above shows the change of yearly crime counts from 2003 to 2018. 

From the line chart we can see that 2015 has the highest cirme rate while 2011 are around the lowest crime rate. The number of crimes descreased gradually from 2003 to 2011, where it reached the lowest in 2011. But from 2012 to 2017, the number of crimes rose rapidly. The crime rate was so high in 2015, which may be relevant to the Propostition 47 signed by the governor in the California referendum in 2014 that led to more theft and robbery crimes.

We also noticed that 2018 data is far less than other years, which is due to incomplete data with just first 4 months in 2018. So we will look at monthly count below.

In [0]:
#check the last record of 2018
last = spark.sql("select Date from sf_crime where Year = 2018 and Month = 5 order by Date desc limit 10")
display(last)

Date 05/15/2018 05/15/2018 05/15/2018 05/15/2018 05/15/2018 05/15/2018 05/15/2018 05/15/2018 05/15/2018 05/15/2018

In [0]:
crime_monthly = spark.sql("select count(*) as monthly_crimes, Year, Month from sf_crime group by 2,3 order by 2,3")
display(crime_monthly)

monthly_crimes Year Month 12366 2003 1 11377 2003 2 12400 2003 3 12251 2003 4 12041 2003 5 11451 2003 6 11889 2003 7 12390 2003 8 12213 2003 9 12475 2003 10 11338 2003 11 10612 2003 12 12473 2004 1 11380 2004 2 12490 2004 3 12190 2004 4 12749 2004 5 11728 2004 6 12118 2004 7 12132 2004 8 11322 2004 9 11663 2004 10 10668 2004 11 11141 2004 12 11887 2005 1 10304 2005 2 11725 2005 3 11567 2005 4 11784 2005 5 11218 2005 6 11491 2005 7 11804 2005 8 11260 2005 9 12269 2005 10 11150 2005 11 10589 2005 12 11349 2006 1 10236 2006 2 11113 2006 3 11062 2006 4 10960 2006 5 11140 2006 6 11259 2006 7 12048 2006 8 10901 2006 9 11280 2006 10 10271 2006 11 10237 2006 12 11062 2007 1 9969 2007 2 11286 2007 3 10524 2007 4 10846 2007 5 10624 2007 6 11437 2007 7 11370 2007 8 11477 2007 9 11632 2007 10 11120 2007 11 10424 2007 12 11397 2008 1 10517 2008 2 11848 2008 3 10894 2008 4 11301 2008 5 10275 2008 6 11875 2008 7 11060 2008 8 11610 2008 9 12376 2008 10 11383 2008 11 10706 2008 12 11466 2009 1 10226 2009 2 12162 2009 3 11608 2009 4 10817 2009 5 9924 2009 6 11043 2009 7 12598 2009 8 11900 2009 9 11503 2009 10 10841 2009 11 10221 2009 12 10805 2010 1 9779 2010 2 10613 2010 3 9995 2010 4 11082 2010 5 10443 2010 6 10521 2010 7 11200 2010 8 11295 2010 9 11382 2010 10 10339 2010 11 10304 2010 12 11058 2011 1 9542 2011 2 10885 2011 3 10636 2011 4 10913 2011 5 10512 2011 6 10294 2011 7 10605 2011 8 10762 2011 9 10899 2011 10 10105 2011 11 10502 2011 12 11109 2012 1 10393 2012 2 10985 2012 3 10788 2012 4 11336 2012 5 10681 2012 6 11160 2012 7 11932 2012 8 11495 2012 9 12692 2012 10 11564 2012 11 11329 2012 12 12311 2013 1 11308 2013 2 12458 2013 3 11988 2013 4 12362 2013 5 11971 2013 6 12659 2013 7 13205 2013 8 13325 2013 9 13387 2013 10 11822 2013 11 10868 2013 12 12373 2014 1 11124 2014 2 12407 2014 3 11908 2014 4 11840 2014 5 11499 2014 6 12281 2014 7 12990 2014 8 12081 2014 9 12896 2014 10 11707 2014 11 11738 2014 12 13181 2015 1 11882 2015 2 13463 2015 3 12526 2015 4 13318 2015 5 12853 2015 6 12949 2015 7 13317 2015 8 12476 2015 9 12697 2015 10 11691 2015 11 11106 2015 12 12532 2016 1 11743 2016 2 11950 2016 3 11905 2016 4 12313 2016 5 11727 2016 6 11832 2016 7 12104 2016 8 12092 2016 9 12913 2016 10 12254 2016 11 12629 2016 12 12687 2017 1 11780 2017 2 13250 2017 3 12452 2017 4 12758 2017 5 12186 2017 6 12717 2017 7 12428 2017 8 12204 2017 9 12970 2017 10 11940 2017 11 12115 2017 12 11667 2018 1 9565 2018 2 10354 2018 3 9954 2018 4 3519 2018 5

#####Analysis
As can be seen from above, 2018 only contains full records for the first 4 months and records only up to 15th in May. Therefore when comparing yearly difference or monthly difference for May-December, 2018 should be left out.

In 2013-2017, the crime rates are above the average of the 15-year range.

**Calendar effect:*

January, March, May, August and October have seen a rise on crimes from previous month, which may have something to do with one more day in calender. 

Feburary have the lowest crime numbers, which is due to the fact that it has 2 or 3 days fewer than other months.

#### 1. number of crimes for different category

In [0]:
crimeCategory = spark.sql("SELECT  category, COUNT(*) AS Count FROM sf_crime GROUP BY category ORDER BY Count DESC")
display(crimeCategory)

category Count LARCENY/THEFT 477975 OTHER OFFENSES 301874 NON-CRIMINAL 236928 ASSAULT 167042 VEHICLE THEFT 126228 DRUG/NARCOTIC 117821 VANDALISM 114718 WARRANTS 99821 BURGLARY 91067 SUSPICIOUS OCC 79087 ROBBERY 54467 MISSING PERSON 44268 FRAUD 41348 FORGERY/COUNTERFEITING 22995 SECONDARY CODES 22378 WEAPON LAWS 21004 TRESPASS 19194 PROSTITUTION 16501 STOLEN PROPERTY 11450 DISORDERLY CONDUCT 9932 DRUNKENNESS 9760 SEX OFFENSES, FORCIBLE 8747 RECOVERED VEHICLE 8688 DRIVING UNDER THE INFLUENCE 5652 KIDNAPPING 4282 ARSON 3875 EMBEZZLEMENT 2984 LIQUOR LAWS 2840 LOITERING 2402 SUICIDE 1291 BAD CHECKS 924 BRIBERY 796 EXTORTION 729 GAMBLING 343 PORNOGRAPHY/OBSCENE MAT 57 SEX OFFENSES, NON FORCIBLE 43 TREA 14

#####Analysis
As can be seen from the bar chart, Larceny/Theft is the most committed crime at SF, which is about 60% more than the second crime 'Other offenses'. 

As mentioned before, Proposition 47, the 2014 ballot measure that reclassified nonviolent thefts as misdemeanors if the stolen goods are worth less than $950, had emboldened thieves.

Other offenses, non-criminal and assault are the second, third and forth of the most committed crimes respectively.

#### 2. the number of crimes for different district

In [0]:
crime_district = spark.sql("SELECT PdDistrict, count(*) as count from sf_crime group by 1 order by 2 desc")
display(crime_district)

PdDistrict count SOUTHERN 390692 MISSION 288985 NORTHERN 266435 CENTRAL 221923 BAYVIEW 205480 TENDERLOIN 186954 INGLESIDE 181092 TARAVAL 155461 PARK 119698 RICHMOND 112804 NA 1

#####Analysis
As can be seen from the bar chart, most crimes are committed in Sounthern district. Mission and Northern district follows as the second and third place with crime counts similar to each other. Richmond and Park district have the lowest crime numbers. 

Therefore, when deciding where to live, I would suggest to consider those districts with lower crime rate such as Richmond, and avoiding areas close to Southern district. If go traveling or commuting to districts with higher crime rate, keep careful watch for potential danger in your surroundings .

#### 3. the number of crimes each Sunday at "SF downtown".   
SF downtown is defiend  via the range of spatial location. For example, you can use a rectangle to define the SF downtown, or you can define a cicle with center as well. 
San Francisco Latitude and longitude coordinates are: 37.773972, -122.431297. X and Y represents each. So we assume SF downtown spacial range: X (-122.4213,-122.4313), Y(37.7540,37.7740).

In [0]:
crime_Sun_SFdowntown = spark.sql("""
                                 with sunday_downtown_crime as(
                                 select substring(Date,1,5) as date,
                                        substring(Date,7) as year
                                 from sf_crime 
                                 where (DayOfWeek = 'Sunday' 
                                 and X <= -122.4213 
                                 and X >= -122.4313 
                                 and Y >= 37.7540 
                                 and Y <= 37.7740)
                                 )
                                 select year, date, count(*) as count
                                 from sunday_downtown_crime
                                 group by 1,2 
                                 order by 1,2
                                 """)

display(crime_Sun_SFdowntown)

year date count 2003 01/05 13 2003 01/12 20 2003 01/19 17 2003 01/26 13 2003 02/02 14 2003 02/09 22 2003 02/16 12 2003 02/23 14 2003 03/02 16 2003 03/09 8 2003 03/16 18 2003 03/23 9 2003 03/30 20 2003 04/06 8 2003 04/13 13 2003 04/20 13 2003 04/27 5 2003 05/04 9 2003 05/11 5 2003 05/18 16 2003 05/25 10 2003 06/01 11 2003 06/08 25 2003 06/15 6 2003 06/22 6 2003 06/29 16 2003 07/06 10 2003 07/13 17 2003 07/20 17 2003 07/27 8 2003 08/03 6 2003 08/10 5 2003 08/17 11 2003 08/24 10 2003 08/31 8 2003 09/07 13 2003 09/14 10 2003 09/21 10 2003 09/28 7 2003 10/05 14 2003 10/12 13 2003 10/19 11 2003 10/26 16 2003 11/02 17 2003 11/09 16 2003 11/16 10 2003 11/23 14 2003 11/30 6 2003 12/07 4 2003 12/14 10 2003 12/21 12 2003 12/28 11 2004 01/04 14 2004 01/11 15 2004 01/18 6 2004 01/25 22 2004 02/01 16 2004 02/08 20 2004 02/15 11 2004 02/22 18 2004 02/29 11 2004 03/07 16 2004 03/14 8 2004 03/21 8 2004 03/28 12 2004 04/04 13 2004 04/11 15 2004 04/18 10 2004 04/25 11 2004 05/02 12 2004 05/09 16 2004 05/16 22 2004 05/23 5 2004 05/30 15 2004 06/06 16 2004 06/13 10 2004 06/20 10 2004 06/27 11 2004 07/04 10 2004 07/11 11 2004 07/18 18 2004 07/25 11 2004 08/01 16 2004 08/08 13 2004 08/15 8 2004 08/22 12 2004 08/29 22 2004 09/05 15 2004 09/12 9 2004 09/19 21 2004 09/26 14 2004 10/03 13 2004 10/10 18 2004 10/17 15 2004 10/24 19 2004 10/31 31 2004 11/07 14 2004 11/14 18 2004 11/21 10 2004 11/28 8 2004 12/05 9 2004 12/12 10 2004 12/19 12 2004 12/26 7 2005 01/02 8 2005 01/09 16 2005 01/16 8 2005 01/23 10 2005 01/30 10 2005 02/06 10 2005 02/13 7 2005 02/20 14 2005 02/27 10 2005 03/06 14 2005 03/13 13 2005 03/20 8 2005 03/27 6 2005 04/03 18 2005 04/10 16 2005 04/17 4 2005 04/24 11 2005 05/01 11 2005 05/08 12 2005 05/15 7 2005 05/22 6 2005 05/29 10 2005 06/05 9 2005 06/12 10 2005 06/19 12 2005 06/26 18 2005 07/03 12 2005 07/10 14 2005 07/17 9 2005 07/24 13 2005 07/31 10 2005 08/07 10 2005 08/14 3 2005 08/21 11 2005 08/28 13 2005 09/04 9 2005 09/11 8 2005 09/18 5 2005 09/25 19 2005 10/02 16 2005 10/09 25 2005 10/16 7 2005 10/23 9 2005 10/30 9 2005 11/06 8 2005 11/13 11 2005 11/20 16 2005 11/27 13 2005 12/04 16 2005 12/11 6 2005 12/18 7 2005 12/25 15 2006 01/01 20 2006 01/08 14 2006 01/15 4 2006 01/22 11 2006 01/29 13 2006 02/05 11 2006 02/12 11 2006 02/19 20 2006 02/26 9 2006 03/05 7 2006 03/12 6 2006 03/19 13 2006 03/26 10 2006 04/02 16 2006 04/09 14 2006 04/16 7 2006 04/23 12 2006 04/30 12 2006 05/07 7 2006 05/14 14 2006 05/21 18 2006 05/28 13 2006 06/04 9 2006 06/11 9 2006 06/18 10 2006 06/25 18 2006 07/02 20 2006 07/09 12 2006 07/16 22 2006 07/23 15 2006 07/30 26 2006 08/06 6 2006 08/13 21 2006 08/20 16 2006 08/27 16 2006 09/03 8 2006 09/10 12 2006 09/17 7 2006 09/24 6 2006 10/01 19 2006 10/08 10 2006 10/15 8 2006 10/22 12 2006 10/29 9 2006 11/05 14 2006 11/12 16 2006 11/19 12 2006 11/26 9 2006 12/03 21 2006 12/10 14 2006 12/17 9 2006 12/24 5 2006 12/31 11 2007 01/07 19 2007 01/14 6 2007 01/21 7 2007 01/28 12 2007 02/04 6 2007 02/11 13 2007 02/18 16 2007 02/25 8 2007 03/04 6 2007 03/11 15 2007 03/18 5 2007 03/25 17 2007 04/01 18 2007 04/08 13 2007 04/15 13 2007 04/22 12 2007 04/29 18 2007 05/06 3 2007 05/13 5 2007 05/20 11 2007 05/27 12 2007 06/03 7 2007 06/10 12 2007 06/17 18 2007 06/24 28 2007 07/01 15 2007 07/08 12 2007 07/15 9 2007 07/22 8 2007 07/29 8 2007 08/05 8 2007 08/12 18 2007 08/19 8 2007 08/26 14 2007 09/02 9 2007 09/09 23 2007 09/16 16 2007 09/23 1 2007 09/30 7 2007 10/07 14 2007 10/14 11 2007 10/21 8 2007 10/28 18 2007 11/04 12 2007 11/11 14 2007 11/18 16 2007 11/25 11 2007 12/02 8 2007 12/09 11 2007 12/23 5 2007 12/30 12 2008 01/06 7 2008 01/13 10 2008 01/20 19 2008 01/27 6 2008 02/03 12 2008 02/10 9 2008 02/17 7 2008 02/24 8 2008 03/02 8 2008 03/09 10 2008 03/16 14 2008 03/23 16 2008 03/30 7 2008 04/06 8 2008 04/13 11 2008 04/20 14 2008 04/27 8 2008 05/04 19 2008 05/11 9 2008 05/18 10 2008 05/25 10 2008 06/01 23 2008 06/08 13 2008 06/15 20 2008 06/22 5 2008 06/29 13 2008 07/06 9 2008 07/13 14 2008 07/20 7 2008 07/27 2 2008 08/03 9 2008 0

#####Analysis
The bar chart above shows the history crime rate on a certain day of week(eg.Sunday) at a customized area (eg.SF downtown).
On 06/30, 2013, there was a peak of 54 crimes that Sunday at SF downtown. Police officers could further analyze that abnormal case to gain possible insights on the reason.

#### 4.the number of crime in each month of 2015, 2016, 2017, 2018.

In [0]:
monthly_crime = spark.sql("""
select Year, Month, count(*) as crime_number from sf_crime
where Year in (2015,2016,2017,2018) group by 1,2 order by 1,2
""")

display(monthly_crime)

Year Month crime_number 2015 1 13181 2015 2 11882 2015 3 13463 2015 4 12526 2015 5 13318 2015 6 12853 2015 7 12949 2015 8 13317 2015 9 12476 2015 10 12697 2015 11 11691 2015 12 11106 2016 1 12532 2016 2 11743 2016 3 11950 2016 4 11905 2016 5 12313 2016 6 11727 2016 7 11832 2016 8 12104 2016 9 12092 2016 10 12913 2016 11 12254 2016 12 12629 2017 1 12687 2017 2 11780 2017 3 13250 2017 4 12452 2017 5 12758 2017 6 12186 2017 7 12717 2017 8 12428 2017 9 12204 2017 10 12970 2017 11 11940 2017 12 12115 2018 1 11667 2018 2 9565 2018 3 10354 2018 4 9954 2018 5 3519

In [0]:
display(monthly_crime)

Year Month crime_number 2015 1 13181 2015 2 11882 2015 3 13463 2015 4 12526 2015 5 13318 2015 6 12853 2015 7 12949 2015 8 13317 2015 9 12476 2015 10 12697 2015 11 11691 2015 12 11106 2016 1 12532 2016 2 11743 2016 3 11950 2016 4 11905 2016 5 12313 2016 6 11727 2016 7 11832 2016 8 12104 2016 9 12092 2016 10 12913 2016 11 12254 2016 12 12629 2017 1 12687 2017 2 11780 2017 3 13250 2017 4 12452 2017 5 12758 2017 6 12186 2017 7 12717 2017 8 12428 2017 9 12204 2017 10 12970 2017 11 11940 2017 12 12115 2018 1 11667 2018 2 9565 2018 3 10354 2018 4 9954 2018 5 3519

#####Analysis

From the bar chart, we can see that in 2015-2017, January, March and May have most crimes.
August in 2015, October and December in 2016 and 2017 also have more crimes.

From the line chart, we can see a descreasing trend of crime rate from fourth quarter of 2015 to October 2016. It was not until the beginning of 2018 that crime rate significantly dropped in the first four months. This might suggest a good job done by the SFPD in 2018.

Why there's always more crimes in certain month besides the fact of more days on calendar? To answer this question, I took a look at crimes in December and January by date.

In [0]:
crime_Dec = spark.sql("select Year, substring(Date,1,5) as date, count(*) as crime_number from sf_crime where Year in (2015,2016,2017,2018) and Month in (12) group by 1,2 order by 1, 2")
display(crime_Dec)

Year date crime_number 2015 12/01 417 2015 12/02 363 2015 12/03 389 2015 12/04 423 2015 12/05 379 2015 12/06 358 2015 12/07 354 2015 12/08 386 2015 12/09 366 2015 12/10 387 2015 12/11 364 2015 12/12 338 2015 12/13 337 2015 12/14 359 2015 12/15 346 2015 12/16 359 2015 12/17 340 2015 12/18 283 2015 12/19 253 2015 12/20 298 2015 12/21 271 2015 12/22 367 2015 12/23 388 2015 12/24 414 2015 12/25 271 2015 12/26 371 2015 12/27 342 2015 12/28 350 2015 12/29 369 2015 12/30 463 2015 12/31 401 2016 12/01 472 2016 12/02 436 2016 12/03 360 2016 12/04 403 2016 12/05 403 2016 12/06 425 2016 12/07 354 2016 12/08 358 2016 12/09 410 2016 12/10 422 2016 12/11 436 2016 12/12 436 2016 12/13 400 2016 12/14 492 2016 12/15 396 2016 12/16 468 2016 12/17 404 2016 12/18 355 2016 12/19 425 2016 12/20 443 2016 12/21 386 2016 12/22 452 2016 12/23 445 2016 12/24 364 2016 12/25 218 2016 12/26 357 2016 12/27 424 2016 12/28 394 2016 12/29 416 2016 12/30 409 2016 12/31 466 2017 12/01 450 2017 12/02 367 2017 12/03 340 2017 12/04 371 2017 12/05 386 2017 12/06 389 2017 12/07 430 2017 12/08 403 2017 12/09 376 2017 12/10 421 2017 12/11 398 2017 12/12 383 2017 12/13 396 2017 12/14 439 2017 12/15 458 2017 12/16 412 2017 12/17 376 2017 12/18 407 2017 12/19 352 2017 12/20 417 2017 12/21 421 2017 12/22 401 2017 12/23 425 2017 12/24 306 2017 12/25 265 2017 12/26 312 2017 12/27 374 2017 12/28 408 2017 12/29 401 2017 12/30 418 2017 12/31 413

In [0]:
crime_Jan = spark.sql("select Year, substring(Date,1,5) as date, count(*) as crime_number from sf_crime where Year in (2015,2016,2017,2018) and Month in (1) group by 1,2 order by 1, 2")
display(crime_Jan)

Year date crime_number 2015 01/01 538 2015 01/02 423 2015 01/03 436 2015 01/04 323 2015 01/05 447 2015 01/06 406 2015 01/07 391 2015 01/08 381 2015 01/09 489 2015 01/10 414 2015 01/11 348 2015 01/12 420 2015 01/13 404 2015 01/14 457 2015 01/15 481 2015 01/16 464 2015 01/17 475 2015 01/18 390 2015 01/19 470 2015 01/20 422 2015 01/21 396 2015 01/22 413 2015 01/23 429 2015 01/24 440 2015 01/25 356 2015 01/26 364 2015 01/27 456 2015 01/28 427 2015 01/29 431 2015 01/30 454 2015 01/31 436 2016 01/01 543 2016 01/02 387 2016 01/03 304 2016 01/04 353 2016 01/05 367 2016 01/06 350 2016 01/07 418 2016 01/08 432 2016 01/09 396 2016 01/10 388 2016 01/11 346 2016 01/12 389 2016 01/13 407 2016 01/14 380 2016 01/15 427 2016 01/16 433 2016 01/17 373 2016 01/18 358 2016 01/19 367 2016 01/20 438 2016 01/21 434 2016 01/22 447 2016 01/23 468 2016 01/24 393 2016 01/25 390 2016 01/26 438 2016 01/27 427 2016 01/28 380 2016 01/29 495 2016 01/30 407 2016 01/31 397 2017 01/01 472 2017 01/02 378 2017 01/03 403 2017 01/04 370 2017 01/05 404 2017 01/06 424 2017 01/07 395 2017 01/08 361 2017 01/09 367 2017 01/10 393 2017 01/11 373 2017 01/12 373 2017 01/13 429 2017 01/14 454 2017 01/15 406 2017 01/16 406 2017 01/17 389 2017 01/18 388 2017 01/19 415 2017 01/20 415 2017 01/21 463 2017 01/22 410 2017 01/23 445 2017 01/24 425 2017 01/25 399 2017 01/26 434 2017 01/27 424 2017 01/28 421 2017 01/29 409 2017 01/30 423 2017 01/31 419 2018 01/01 435 2018 01/02 353 2018 01/03 382 2018 01/04 411 2018 01/05 423 2018 01/06 405 2018 01/07 391 2018 01/08 333 2018 01/09 370 2018 01/10 387 2018 01/11 356 2018 01/12 415 2018 01/13 398 2018 01/14 370 2018 01/15 372 2018 01/16 346 2018 01/17 348 2018 01/18 392 2018 01/19 405 2018 01/20 345 2018 01/21 332 2018 01/22 377 2018 01/23 396 2018 01/24 358 2018 01/25 375 2018 01/26 377 2018 01/27 393 2018 01/28 338 2018 01/29 358 2018 01/30 369 2018 01/31 357

#####Analysis
From the line charts, it seems that days before New Year's Day always has the most crimes, but would drop immediatly after New Year's day. There's also the lowest crimes committed on the exact day of Chritmas. The period after Christmas until New Year's Day has witnessed a growing number of crimes.

Therefore, during these periods, the police should arrange more force, despite the fact that it should be holiday off. 

(Or is it possible that just because most police officers take breaks during this period and thus reduce the police force, that the criminals become more active and commit more crimes?)

#### 5.travel suggestion to visit SF
Analyze the number of crime w.r.t the hour in certian day like 2015/6/15, 2016/6/15, 2017/6/15.

In [0]:
hourly_crime_certainDay = spark.sql("""
select Date, 
Hour, 
count(*) as crime_number, 
Year
from sf_crime where Date like '06/15/%'
group by 4,1,2 order by 1,2
""")
display(hourly_crime_certainDay)

Date Hour crime_number Year 06/15/2003 0 24 2003 06/15/2003 1 15 2003 06/15/2003 2 16 2003 06/15/2003 3 6 2003 06/15/2003 4 3 2003 06/15/2003 5 1 2003 06/15/2003 6 3 2003 06/15/2003 7 9 2003 06/15/2003 8 5 2003 06/15/2003 9 10 2003 06/15/2003 10 13 2003 06/15/2003 11 9 2003 06/15/2003 12 12 2003 06/15/2003 13 8 2003 06/15/2003 14 16 2003 06/15/2003 15 12 2003 06/15/2003 16 12 2003 06/15/2003 17 17 2003 06/15/2003 18 21 2003 06/15/2003 19 21 2003 06/15/2003 20 19 2003 06/15/2003 21 13 2003 06/15/2003 22 26 2003 06/15/2003 23 13 2003 06/15/2004 0 16 2004 06/15/2004 1 8 2004 06/15/2004 2 1 2004 06/15/2004 3 4 2004 06/15/2004 4 4 2004 06/15/2004 5 5 2004 06/15/2004 6 6 2004 06/15/2004 7 3 2004 06/15/2004 8 19 2004 06/15/2004 9 31 2004 06/15/2004 10 25 2004 06/15/2004 11 17 2004 06/15/2004 12 25 2004 06/15/2004 13 17 2004 06/15/2004 14 20 2004 06/15/2004 15 22 2004 06/15/2004 16 22 2004 06/15/2004 17 23 2004 06/15/2004 18 26 2004 06/15/2004 19 24 2004 06/15/2004 20 20 2004 06/15/2004 21 21 2004 06/15/2004 22 22 2004 06/15/2004 23 18 2004 06/15/2005 0 20 2005 06/15/2005 1 10 2005 06/15/2005 2 5 2005 06/15/2005 4 3 2005 06/15/2005 5 9 2005 06/15/2005 6 6 2005 06/15/2005 7 6 2005 06/15/2005 8 19 2005 06/15/2005 9 9 2005 06/15/2005 10 16 2005 06/15/2005 11 20 2005 06/15/2005 12 21 2005 06/15/2005 13 8 2005 06/15/2005 14 17 2005 06/15/2005 15 13 2005 06/15/2005 16 18 2005 06/15/2005 17 19 2005 06/15/2005 18 32 2005 06/15/2005 19 19 2005 06/15/2005 20 23 2005 06/15/2005 21 12 2005 06/15/2005 22 18 2005 06/15/2005 23 12 2005 06/15/2006 0 19 2006 06/15/2006 1 5 2006 06/15/2006 2 10 2006 06/15/2006 3 5 2006 06/15/2006 4 7 2006 06/15/2006 5 2 2006 06/15/2006 6 8 2006 06/15/2006 7 8 2006 06/15/2006 8 21 2006 06/15/2006 9 17 2006 06/15/2006 10 22 2006 06/15/2006 11 14 2006 06/15/2006 12 28 2006 06/15/2006 13 20 2006 06/15/2006 14 20 2006 06/15/2006 15 31 2006 06/15/2006 16 31 2006 06/15/2006 17 13 2006 06/15/2006 18 18 2006 06/15/2006 19 21 2006 06/15/2006 20 20 2006 06/15/2006 21 21 2006 06/15/2006 22 18 2006 06/15/2006 23 16 2006 06/15/2007 0 26 2007 06/15/2007 1 6 2007 06/15/2007 2 8 2007 06/15/2007 3 8 2007 06/15/2007 4 3 2007 06/15/2007 5 6 2007 06/15/2007 6 9 2007 06/15/2007 7 12 2007 06/15/2007 8 17 2007 06/15/2007 9 14 2007 06/15/2007 10 23 2007 06/15/2007 11 16 2007 06/15/2007 12 34 2007 06/15/2007 13 19 2007 06/15/2007 14 19 2007 06/15/2007 15 29 2007 06/15/2007 16 29 2007 06/15/2007 17 22 2007 06/15/2007 18 24 2007 06/15/2007 19 31 2007 06/15/2007 20 23 2007 06/15/2007 21 25 2007 06/15/2007 22 21 2007 06/15/2007 23 21 2007 06/15/2008 0 32 2008 06/15/2008 1 14 2008 06/15/2008 2 14 2008 06/15/2008 3 5 2008 06/15/2008 4 3 2008 06/15/2008 5 4 2008 06/15/2008 6 6 2008 06/15/2008 7 7 2008 06/15/2008 8 10 2008 06/15/2008 9 10 2008 06/15/2008 10 13 2008 06/15/2008 11 18 2008 06/15/2008 12 21 2008 06/15/2008 13 19 2008 06/15/2008 14 14 2008 06/15/2008 15 23 2008 06/15/2008 16 17 2008 06/15/2008 17 25 2008 06/15/2008 18 17 2008 06/15/2008 19 14 2008 06/15/2008 20 15 2008 06/15/2008 21 14 2008 06/15/2008 22 18 2008 06/15/2008 23 9 2008 06/15/2009 0 21 2009 06/15/2009 1 7 2009 06/15/2009 2 15 2009 06/15/2009 3 4 2009 06/15/2009 4 7 2009 06/15/2009 5 2 2009 06/15/2009 6 6 2009 06/15/2009 7 8 2009 06/15/2009 8 13 2009 06/15/2009 9 20 2009 06/15/2009 10 18 2009 06/15/2009 11 18 2009 06/15/2009 12 30 2009 06/15/2009 13 24 2009 06/15/2009 14 25 2009 06/15/2009 15 19 2009 06/15/2009 16 19 2009 06/15/2009 17 26 2009 06/15/2009 18 22 2009 06/15/2009 19 20 2009 06/15/2009 20 10 2009 06/15/2009 21 13 2009 06/15/2009 22 13 2009 06/15/2009 23 13 2009 06/15/2010 0 15 2010 06/15/2010 1 5 2010 06/15/2010 2 9 2010 06/15/2010 3 3 2010 06/15/2010 4 2 2010 06/15/2010 6 6 2010 06/15/2010 7 12 2010 06/15/2010 8 11 2010 06/15/2010 9 15 2010 06/15/2010 10 15 2010 06/15/2010 11 18 2010 06/15/2010 12 25 2010 06/15/2010 13 20 2010 06/15/2010 14 24 2010 06/15/2010 15 36 2010 06/15/2010 16 17 2010 06/15/2010 17 20 2010 06/15/2010 18 23 2010 06/15/2010 19 28 2010 06/15/201

In [0]:
#plot by total number
display(hourly_crime_certainDay)

Date Hour crime_number Year 06/15/2003 0 24 2003 06/15/2003 1 15 2003 06/15/2003 2 16 2003 06/15/2003 3 6 2003 06/15/2003 4 3 2003 06/15/2003 5 1 2003 06/15/2003 6 3 2003 06/15/2003 7 9 2003 06/15/2003 8 5 2003 06/15/2003 9 10 2003 06/15/2003 10 13 2003 06/15/2003 11 9 2003 06/15/2003 12 12 2003 06/15/2003 13 8 2003 06/15/2003 14 16 2003 06/15/2003 15 12 2003 06/15/2003 16 12 2003 06/15/2003 17 17 2003 06/15/2003 18 21 2003 06/15/2003 19 21 2003 06/15/2003 20 19 2003 06/15/2003 21 13 2003 06/15/2003 22 26 2003 06/15/2003 23 13 2003 06/15/2004 0 16 2004 06/15/2004 1 8 2004 06/15/2004 2 1 2004 06/15/2004 3 4 2004 06/15/2004 4 4 2004 06/15/2004 5 5 2004 06/15/2004 6 6 2004 06/15/2004 7 3 2004 06/15/2004 8 19 2004 06/15/2004 9 31 2004 06/15/2004 10 25 2004 06/15/2004 11 17 2004 06/15/2004 12 25 2004 06/15/2004 13 17 2004 06/15/2004 14 20 2004 06/15/2004 15 22 2004 06/15/2004 16 22 2004 06/15/2004 17 23 2004 06/15/2004 18 26 2004 06/15/2004 19 24 2004 06/15/2004 20 20 2004 06/15/2004 21 21 2004 06/15/2004 22 22 2004 06/15/2004 23 18 2004 06/15/2005 0 20 2005 06/15/2005 1 10 2005 06/15/2005 2 5 2005 06/15/2005 4 3 2005 06/15/2005 5 9 2005 06/15/2005 6 6 2005 06/15/2005 7 6 2005 06/15/2005 8 19 2005 06/15/2005 9 9 2005 06/15/2005 10 16 2005 06/15/2005 11 20 2005 06/15/2005 12 21 2005 06/15/2005 13 8 2005 06/15/2005 14 17 2005 06/15/2005 15 13 2005 06/15/2005 16 18 2005 06/15/2005 17 19 2005 06/15/2005 18 32 2005 06/15/2005 19 19 2005 06/15/2005 20 23 2005 06/15/2005 21 12 2005 06/15/2005 22 18 2005 06/15/2005 23 12 2005 06/15/2006 0 19 2006 06/15/2006 1 5 2006 06/15/2006 2 10 2006 06/15/2006 3 5 2006 06/15/2006 4 7 2006 06/15/2006 5 2 2006 06/15/2006 6 8 2006 06/15/2006 7 8 2006 06/15/2006 8 21 2006 06/15/2006 9 17 2006 06/15/2006 10 22 2006 06/15/2006 11 14 2006 06/15/2006 12 28 2006 06/15/2006 13 20 2006 06/15/2006 14 20 2006 06/15/2006 15 31 2006 06/15/2006 16 31 2006 06/15/2006 17 13 2006 06/15/2006 18 18 2006 06/15/2006 19 21 2006 06/15/2006 20 20 2006 06/15/2006 21 21 2006 06/15/2006 22 18 2006 06/15/2006 23 16 2006 06/15/2007 0 26 2007 06/15/2007 1 6 2007 06/15/2007 2 8 2007 06/15/2007 3 8 2007 06/15/2007 4 3 2007 06/15/2007 5 6 2007 06/15/2007 6 9 2007 06/15/2007 7 12 2007 06/15/2007 8 17 2007 06/15/2007 9 14 2007 06/15/2007 10 23 2007 06/15/2007 11 16 2007 06/15/2007 12 34 2007 06/15/2007 13 19 2007 06/15/2007 14 19 2007 06/15/2007 15 29 2007 06/15/2007 16 29 2007 06/15/2007 17 22 2007 06/15/2007 18 24 2007 06/15/2007 19 31 2007 06/15/2007 20 23 2007 06/15/2007 21 25 2007 06/15/2007 22 21 2007 06/15/2007 23 21 2007 06/15/2008 0 32 2008 06/15/2008 1 14 2008 06/15/2008 2 14 2008 06/15/2008 3 5 2008 06/15/2008 4 3 2008 06/15/2008 5 4 2008 06/15/2008 6 6 2008 06/15/2008 7 7 2008 06/15/2008 8 10 2008 06/15/2008 9 10 2008 06/15/2008 10 13 2008 06/15/2008 11 18 2008 06/15/2008 12 21 2008 06/15/2008 13 19 2008 06/15/2008 14 14 2008 06/15/2008 15 23 2008 06/15/2008 16 17 2008 06/15/2008 17 25 2008 06/15/2008 18 17 2008 06/15/2008 19 14 2008 06/15/2008 20 15 2008 06/15/2008 21 14 2008 06/15/2008 22 18 2008 06/15/2008 23 9 2008 06/15/2009 0 21 2009 06/15/2009 1 7 2009 06/15/2009 2 15 2009 06/15/2009 3 4 2009 06/15/2009 4 7 2009 06/15/2009 5 2 2009 06/15/2009 6 6 2009 06/15/2009 7 8 2009 06/15/2009 8 13 2009 06/15/2009 9 20 2009 06/15/2009 10 18 2009 06/15/2009 11 18 2009 06/15/2009 12 30 2009 06/15/2009 13 24 2009 06/15/2009 14 25 2009 06/15/2009 15 19 2009 06/15/2009 16 19 2009 06/15/2009 17 26 2009 06/15/2009 18 22 2009 06/15/2009 19 20 2009 06/15/2009 20 10 2009 06/15/2009 21 13 2009 06/15/2009 22 13 2009 06/15/2009 23 13 2009 06/15/2010 0 15 2010 06/15/2010 1 5 2010 06/15/2010 2 9 2010 06/15/2010 3 3 2010 06/15/2010 4 2 2010 06/15/2010 6 6 2010 06/15/2010 7 12 2010 06/15/2010 8 11 2010 06/15/2010 9 15 2010 06/15/2010 10 15 2010 06/15/2010 11 18 2010 06/15/2010 12 25 2010 06/15/2010 13 20 2010 06/15/2010 14 24 2010 06/15/2010 15 36 2010 06/15/2010 16 17 2010 06/15/2010 17 20 2010 06/15/2010 18 23 2010 06/15/2010 19 28 2010 06/15/201

#####Analysis
The first line chart shows hourly crime rate grouped by years, and the second line chart shows the aggregate count across the 14 years from 2003 to 2017.

From 2003-2017 counts in total, we can see that on June 15 the highest crime number occurs around 12pm, and there's an increasing trend starting from 1pm and reached the second peak around 3pm. 

There're also fewer crimes in the morning than in the afternoon or evening. 

So tourists should better travel in the morning, and should be careful after the lunch time.

A further thinking: what type of crime should they watch out for? Is there a difference of most frequent types of crime across different hours in a day? 

Here I plot line chart for each hour showing hourly trends of different types of crime.

In [0]:
hourly_type_certainD = spark.sql("""
select Year, Hour, Date, Category, count(*) as Count
from sf_crime
where Date like '06/15/%'
group by 1,2,3,4 
order by 1,2,4""")
display(hourly_type_certainD)

Year Hour Date Category Count 2003 0 06/15/2003 ASSAULT 1 2003 0 06/15/2003 DISORDERLY CONDUCT 1 2003 0 06/15/2003 DRUNKENNESS 1 2003 0 06/15/2003 FRAUD 1 2003 0 06/15/2003 LARCENY/THEFT 5 2003 0 06/15/2003 NON-CRIMINAL 1 2003 0 06/15/2003 OTHER OFFENSES 5 2003 0 06/15/2003 SECONDARY CODES 1 2003 0 06/15/2003 STOLEN PROPERTY 1 2003 0 06/15/2003 VANDALISM 1 2003 0 06/15/2003 VEHICLE THEFT 4 2003 0 06/15/2003 WARRANTS 2 2003 1 06/15/2003 DRUG/NARCOTIC 3 2003 1 06/15/2003 LARCENY/THEFT 1 2003 1 06/15/2003 NON-CRIMINAL 1 2003 1 06/15/2003 OTHER OFFENSES 4 2003 1 06/15/2003 VEHICLE THEFT 4 2003 1 06/15/2003 WARRANTS 2 2003 2 06/15/2003 ASSAULT 2 2003 2 06/15/2003 DRUG/NARCOTIC 1 2003 2 06/15/2003 NON-CRIMINAL 2 2003 2 06/15/2003 OTHER OFFENSES 2 2003 2 06/15/2003 SEX OFFENSES, FORCIBLE 1 2003 2 06/15/2003 SUSPICIOUS OCC 1 2003 2 06/15/2003 VANDALISM 2 2003 2 06/15/2003 VEHICLE THEFT 3 2003 2 06/15/2003 WARRANTS 2 2003 3 06/15/2003 ASSAULT 1 2003 3 06/15/2003 DRUG/NARCOTIC 1 2003 3 06/15/2003 NON-CRIMINAL 1 2003 3 06/15/2003 SUSPICIOUS OCC 1 2003 3 06/15/2003 VANDALISM 1 2003 3 06/15/2003 WARRANTS 1 2003 4 06/15/2003 DRUG/NARCOTIC 1 2003 4 06/15/2003 OTHER OFFENSES 2 2003 5 06/15/2003 VANDALISM 1 2003 6 06/15/2003 OTHER OFFENSES 1 2003 6 06/15/2003 VEHICLE THEFT 2 2003 7 06/15/2003 DRUG/NARCOTIC 1 2003 7 06/15/2003 FRAUD 1 2003 7 06/15/2003 LARCENY/THEFT 2 2003 7 06/15/2003 LOITERING 1 2003 7 06/15/2003 MISSING PERSON 1 2003 7 06/15/2003 ROBBERY 1 2003 7 06/15/2003 VANDALISM 1 2003 7 06/15/2003 WARRANTS 1 2003 8 06/15/2003 DRUG/NARCOTIC 2 2003 8 06/15/2003 LARCENY/THEFT 2 2003 8 06/15/2003 WARRANTS 1 2003 9 06/15/2003 ASSAULT 1 2003 9 06/15/2003 DRUG/NARCOTIC 1 2003 9 06/15/2003 FRAUD 1 2003 9 06/15/2003 LARCENY/THEFT 1 2003 9 06/15/2003 NON-CRIMINAL 1 2003 9 06/15/2003 OTHER OFFENSES 2 2003 9 06/15/2003 VANDALISM 1 2003 9 06/15/2003 WARRANTS 1 2003 9 06/15/2003 WEAPON LAWS 1 2003 10 06/15/2003 ASSAULT 1 2003 10 06/15/2003 DRUG/NARCOTIC 1 2003 10 06/15/2003 DRUNKENNESS 1 2003 10 06/15/2003 LARCENY/THEFT 2 2003 10 06/15/2003 LOITERING 1 2003 10 06/15/2003 NON-CRIMINAL 4 2003 10 06/15/2003 OTHER OFFENSES 1 2003 10 06/15/2003 SUSPICIOUS OCC 2 2003 11 06/15/2003 ASSAULT 1 2003 11 06/15/2003 FRAUD 1 2003 11 06/15/2003 LARCENY/THEFT 3 2003 11 06/15/2003 NON-CRIMINAL 1 2003 11 06/15/2003 ROBBERY 1 2003 11 06/15/2003 VANDALISM 2 2003 12 06/15/2003 ASSAULT 2 2003 12 06/15/2003 BURGLARY 2 2003 12 06/15/2003 DRUG/NARCOTIC 1 2003 12 06/15/2003 FRAUD 1 2003 12 06/15/2003 GAMBLING 1 2003 12 06/15/2003 LARCENY/THEFT 1 2003 12 06/15/2003 OTHER OFFENSES 1 2003 12 06/15/2003 VEHICLE THEFT 2 2003 12 06/15/2003 WARRANTS 1 2003 13 06/15/2003 ASSAULT 1 2003 13 06/15/2003 LARCENY/THEFT 2 2003 13 06/15/2003 NON-CRIMINAL 2 2003 13 06/15/2003 SUSPICIOUS OCC 1 2003 13 06/15/2003 VEHICLE THEFT 2 2003 14 06/15/2003 ASSAULT 1 2003 14 06/15/2003 DRUG/NARCOTIC 1 2003 14 06/15/2003 LARCENY/THEFT 5 2003 14 06/15/2003 NON-CRIMINAL 3 2003 14 06/15/2003 OTHER OFFENSES 2 2003 14 06/15/2003 ROBBERY 1 2003 14 06/15/2003 VEHICLE THEFT 1 2003 14 06/15/2003 WARRANTS 1 2003 14 06/15/2003 WEAPON LAWS 1 2003 15 06/15/2003 ASSAULT 1 2003 15 06/15/2003 BURGLARY 1 2003 15 06/15/2003 FRAUD 1 2003 15 06/15/2003 LARCENY/THEFT 5 2003 15 06/15/2003 MISSING PERSON 2 2003 15 06/15/2003 SUSPICIOUS OCC 1 2003 15 06/15/2003 WARRANTS 1 2003 16 06/15/2003 ASSAULT 1 2003 16 06/15/2003 LARCENY/THEFT 2 2003 16 06/15/2003 MISSING PERSON 1 2003 16 06/15/2003 NON-CRIMINAL 2 2003 16 06/15/2003 OTHER OFFENSES 1 2003 16 06/15/2003 SECONDARY CODES 1 2003 16 06/15/2003 SUSPICIOUS OCC 1 2003 16 06/15/2003 VANDALISM 1 2003 16 06/15/2003 VEHICLE THEFT 2 2003 17 06/15/2003 ARSON 1 2003 17 06/15/2003 ASSAULT 3 2003 17 06/15/2003 BURGLARY 1 2003 17 06/15/2003 DRUG/NARCOTIC 1 2003 17 06/15/2003 LARCENY/THEFT 6 2003 17 06/15/2003 NON-CRIMINAL 1 2003 17 06/15/2003 VEHICLE THEFT 1 2003 17 06/15/2003 WARRANTS 3 2003 18 06/15/2003 DRIVING UNDER THE INFLUENCE 1 2003 18 06/15/2003 DRUG/NARCOTIC 1 2003 18 06/15/2003 D

#####Analysis

From the above chart we can see that Larceny/Theft is the most frequent crime from around 11am to 23pm, while Other Offenses is the most frequent crime from 0am to 10am. Therefore, if tourists go out at noon, they should be most careful for their belongings.

Notice that not only tourists should be aware of such crimes, but also shop owners as well. Shoplifting epidemic has worsened over the years, which cause so much heavy burden on shop owner that more and more of the shops has closed because the scale of thefts had made business untenable. According to The New York Times, this spikes in organized retail crime is only witnessed in San Franciso but not other cities in the state, which implies the problem lies not in California law but mostlly on SFPD. The police should definitely provide support for the shop owners and other local business.

#### 6.advice to distribute the police
(1) Step1: Find out the top-3 danger disrict  
(2) Step2: find out the crime event w.r.t category and time (hour) from the result of step 1

In [0]:
#step1:
top3_dangerDistrict = spark.sql("select PdDistrict, count(*) as crime_number from sf_crime group by 1 order by 2 desc limit 3")
display(top3_dangerDistrict)

PdDistrict crime_number SOUTHERN 390692 MISSION 288985 NORTHERN 266435

In [0]:
#step2:
crime_event = spark.sql("""
select PdDistrict, Category, Hour , count(*) as count
from sf_crime 
where PdDistrict in ('SOUTHERN', 'MISSION', 'NORTHERN')
group by 1, 2, 3
order by 1, 2, 3
""")
display(crime_event)

PdDistrict Category Hour count MISSION ARSON 0 35 MISSION ARSON 1 23 MISSION ARSON 2 28 MISSION ARSON 3 36 MISSION ARSON 4 27 MISSION ARSON 5 26 MISSION ARSON 6 12 MISSION ARSON 7 8 MISSION ARSON 8 7 MISSION ARSON 9 14 MISSION ARSON 10 12 MISSION ARSON 11 5 MISSION ARSON 12 13 MISSION ARSON 13 12 MISSION ARSON 14 13 MISSION ARSON 15 17 MISSION ARSON 16 16 MISSION ARSON 17 12 MISSION ARSON 18 19 MISSION ARSON 19 12 MISSION ARSON 20 24 MISSION ARSON 21 24 MISSION ARSON 22 16 MISSION ARSON 23 19 MISSION ASSAULT 0 1570 MISSION ASSAULT 1 1483 MISSION ASSAULT 2 1258 MISSION ASSAULT 3 553 MISSION ASSAULT 4 410 MISSION ASSAULT 5 272 MISSION ASSAULT 6 339 MISSION ASSAULT 7 485 MISSION ASSAULT 8 726 MISSION ASSAULT 9 870 MISSION ASSAULT 10 902 MISSION ASSAULT 11 961 MISSION ASSAULT 12 1175 MISSION ASSAULT 13 1140 MISSION ASSAULT 14 1175 MISSION ASSAULT 15 1279 MISSION ASSAULT 16 1306 MISSION ASSAULT 17 1341 MISSION ASSAULT 18 1349 MISSION ASSAULT 19 1289 MISSION ASSAULT 20 1346 MISSION ASSAULT 21 1338 MISSION ASSAULT 22 1388 MISSION ASSAULT 23 1357 MISSION BAD CHECKS 0 13 MISSION BAD CHECKS 1 1 MISSION BAD CHECKS 2 1 MISSION BAD CHECKS 6 1 MISSION BAD CHECKS 7 1 MISSION BAD CHECKS 8 12 MISSION BAD CHECKS 9 5 MISSION BAD CHECKS 10 4 MISSION BAD CHECKS 11 4 MISSION BAD CHECKS 12 19 MISSION BAD CHECKS 13 2 MISSION BAD CHECKS 14 5 MISSION BAD CHECKS 15 6 MISSION BAD CHECKS 16 3 MISSION BAD CHECKS 17 8 MISSION BAD CHECKS 18 1 MISSION BAD CHECKS 19 4 MISSION BAD CHECKS 20 1 MISSION BAD CHECKS 23 1 MISSION BRIBERY 0 9 MISSION BRIBERY 1 5 MISSION BRIBERY 2 7 MISSION BRIBERY 3 5 MISSION BRIBERY 4 2 MISSION BRIBERY 5 2 MISSION BRIBERY 6 1 MISSION BRIBERY 7 3 MISSION BRIBERY 8 2 MISSION BRIBERY 9 6 MISSION BRIBERY 10 7 MISSION BRIBERY 11 8 MISSION BRIBERY 12 7 MISSION BRIBERY 13 6 MISSION BRIBERY 14 7 MISSION BRIBERY 15 4 MISSION BRIBERY 16 6 MISSION BRIBERY 17 13 MISSION BRIBERY 18 6 MISSION BRIBERY 19 6 MISSION BRIBERY 20 9 MISSION BRIBERY 21 12 MISSION BRIBERY 22 9 MISSION BRIBERY 23 4 MISSION BURGLARY 0 481 MISSION BURGLARY 1 301 MISSION BURGLARY 2 315 MISSION BURGLARY 3 371 MISSION BURGLARY 4 310 MISSION BURGLARY 5 255 MISSION BURGLARY 6 223 MISSION BURGLARY 7 269 MISSION BURGLARY 8 493 MISSION BURGLARY 9 401 MISSION BURGLARY 10 325 MISSION BURGLARY 11 317 MISSION BURGLARY 12 416 MISSION BURGLARY 13 281 MISSION BURGLARY 14 335 MISSION BURGLARY 15 354 MISSION BURGLARY 16 489 MISSION BURGLARY 17 610 MISSION BURGLARY 18 639 MISSION BURGLARY 19 538 MISSION BURGLARY 20 498 MISSION BURGLARY 21 431 MISSION BURGLARY 22 511 MISSION BURGLARY 23 442 MISSION DISORDERLY CONDUCT 0 188 MISSION DISORDERLY CONDUCT 1 129 MISSION DISORDERLY CONDUCT 2 96 MISSION DISORDERLY CONDUCT 3 53 MISSION DISORDERLY CONDUCT 4 43 MISSION DISORDERLY CONDUCT 5 53 MISSION DISORDERLY CONDUCT 6 193 MISSION DISORDERLY CONDUCT 7 199 MISSION DISORDERLY CONDUCT 8 155 MISSION DISORDERLY CONDUCT 9 130 MISSION DISORDERLY CONDUCT 10 122 MISSION DISORDERLY CONDUCT 11 117 MISSION DISORDERLY CONDUCT 12 90 MISSION DISORDERLY CONDUCT 13 84 MISSION DISORDERLY CONDUCT 14 75 MISSION DISORDERLY CONDUCT 15 64 MISSION DISORDERLY CONDUCT 16 64 MISSION DISORDERLY CONDUCT 17 67 MISSION DISORDERLY CONDUCT 18 92 MISSION DISORDERLY CONDUCT 19 83 MISSION DISORDERLY CONDUCT 20 82 MISSION DISORDERLY CONDUCT 21 96 MISSION DISORDERLY CONDUCT 22 126 MISSION DISORDERLY CONDUCT 23 181 MISSION DRIVING UNDER THE INFLUENCE 0 115 MISSION DRIVING UNDER THE INFLUENCE 1 93 MISSION DRIVING UNDER THE INFLUENCE 2 80 MISSION DRIVING UNDER THE INFLUENCE 3 36 MISSION DRIVING UNDER THE INFLUENCE 4 16 MISSION DRIVING UNDER THE INFLUENCE 5 18 MISSION DRIVING UNDER THE INFLUENCE 6 10 MISSION DRIVING UNDER THE INFLUENCE 7 9 MISSION DRIVING UNDER THE INFLUENCE 8 6 MISSION DRIVING UNDER THE INFLUENCE 9 7 MISSION DRIVING UNDER THE INFLUENCE 10 5 MISSION DRIVING UNDER THE INFLUENCE 11 10 MISSION DRIVING UNDER THE INFLUENCE 12 11 MISSION DRIVING UNDER THE INFLUENCE 13 11 MISSION DRIVING UNDER THE INFLUENCE 14 13 MISSION DRIVING UNDER 

#####Analysis

The top 3 dangerous districts are Southern, Mission and Northern.

(See the plot in Tableau)
https://public.tableau.com/profile/zishan.cheng#!/vizhome/sf_crimetop3districthourlycrimecategory/Sheet1?publish=yes

Here I assume that SF police force is arranged as a whole and each district also has some of its own force, so we need to distribute force across districts for SFPD, as well as across hours for each district's police office.

From the stacked bar chart in Tableau we can see that the highest number of crimes occurs at 6pm and 12pm. Additionally, the 5 most committed crimes in each district are larceny/theft, other offenses, non-criminal, assault and durg/narcotic, which accounts for over half of the crimes.


In the Southern district, the period from 11am to 12am has crime number above average of a day within that district. In Mission district, the period is from 11am to 12am. In Northern district, the period is from 12pm to 12am. So more police force should be distributed to these districts during these time periods.

#### 7. hints to adjust policy on different type of crime
For different category of crime, find the percentage of resolution.

In [0]:
%sql select distinct(resolution) as resolve from sf_crime

resolve EXCEPTIONAL CLEARANCE ARREST, BOOKED PROSECUTED FOR LESSER OFFENSE LOCATED UNFOUNDED DISTRICT ATTORNEY REFUSES TO PROSECUTE PSYCHOPATHIC CASE COMPLAINANT REFUSES TO PROSECUTE ARREST, CITED PROSECUTED BY OUTSIDE AGENCY NONE NOT PROSECUTED

In [0]:
crime_total = spark.sql("select Category, count(*) as total from sf_crime group by Category order by 2 desc")
crime_total.createOrReplaceTempView("crime_total")
display(crime_total)
crime_total.cache()

Category total LARCENY/THEFT 477975 OTHER OFFENSES 301874 NON-CRIMINAL 236928 ASSAULT 167042 VEHICLE THEFT 126228 DRUG/NARCOTIC 117821 VANDALISM 114718 WARRANTS 99821 BURGLARY 91067 SUSPICIOUS OCC 79087 ROBBERY 54467 MISSING PERSON 44268 FRAUD 41348 FORGERY/COUNTERFEITING 22995 SECONDARY CODES 22378 WEAPON LAWS 21004 TRESPASS 19194 PROSTITUTION 16501 STOLEN PROPERTY 11450 DISORDERLY CONDUCT 9932 DRUNKENNESS 9760 SEX OFFENSES, FORCIBLE 8747 RECOVERED VEHICLE 8688 DRIVING UNDER THE INFLUENCE 5652 KIDNAPPING 4282 ARSON 3875 EMBEZZLEMENT 2984 LIQUOR LAWS 2840 LOITERING 2402 SUICIDE 1291 BAD CHECKS 924 BRIBERY 796 EXTORTION 729 GAMBLING 343 PORNOGRAPHY/OBSCENE MAT 57 SEX OFFENSES, NON FORCIBLE 43 TREA 14

Out[47]: DataFrame[Category: string, total: bigint]

In [0]:
crime_res = spark.sql("select Category, count(*) as resolved from sf_crime where Resolution != 'NONE' group by 1")
crime_res.createOrReplaceTempView("crime_res")
display(crime_res)

Category resolved FRAUD 9291 SUICIDE 309 LIQUOR LAWS 2527 SECONDARY CODES 9542 MISSING PERSON 22638 OTHER OFFENSES 215059 DRIVING UNDER THE INFLUENCE 5335 WARRANTS 94339 ARSON 701 FORGERY/COUNTERFEITING 8546 GAMBLING 229 BRIBERY 458 ASSAULT 62499 DRUNKENNESS 8035 EXTORTION 152 TREA 8 WEAPON LAWS 14934 LOITERING 2103 SUSPICIOUS OCC 8573 ROBBERY 11385 SEX OFFENSES, FORCIBLE 3172 PROSTITUTION 15651 EMBEZZLEMENT 837 BAD CHECKS 180 DISORDERLY CONDUCT 6474 RECOVERED VEHICLE 575 VANDALISM 12940 DRUG/NARCOTIC 107565 PORNOGRAPHY/OBSCENE MAT 24 TRESPASS 12907 VEHICLE THEFT 10248 NON-CRIMINAL 52553 STOLEN PROPERTY 10011 LARCENY/THEFT 40048 KIDNAPPING 2489 BURGLARY 14414 SEX OFFENSES, NON FORCIBLE 16

In [0]:
resolve_rate = spark.sql("select a.Category, resolved, total, resolved/total as percentage from crime_total a left join crime_res b on a.Category = b.Category order by percentage desc")
display(resolve_rate)

Category resolved total percentage PROSTITUTION 15651 16501 0.9484879704260348 WARRANTS 94339 99821 0.9450816962362629 DRIVING UNDER THE INFLUENCE 5335 5652 0.9439136588818118 DRUG/NARCOTIC 107565 117821 0.912952699433887 LIQUOR LAWS 2527 2840 0.8897887323943662 LOITERING 2103 2402 0.8755203996669442 STOLEN PROPERTY 10011 11450 0.8743231441048035 DRUNKENNESS 8035 9760 0.8232581967213115 OTHER OFFENSES 215059 301874 0.7124131260062145 WEAPON LAWS 14934 21004 0.711007427156732 TRESPASS 12907 19194 0.6724497238720434 GAMBLING 229 343 0.6676384839650146 DISORDERLY CONDUCT 6474 9932 0.6518324607329843 KIDNAPPING 2489 4282 0.5812704343764596 BRIBERY 458 796 0.5753768844221105 TREA 8 14 0.5714285714285714 MISSING PERSON 22638 44268 0.5113851992409867 SECONDARY CODES 9542 22378 0.42640092948431496 PORNOGRAPHY/OBSCENE MAT 24 57 0.42105263157894735 ASSAULT 62499 167042 0.3741514110223776 SEX OFFENSES, NON FORCIBLE 16 43 0.37209302325581395 FORGERY/COUNTERFEITING 8546 22995 0.3716460100021744 SEX OFFENSES, FORCIBLE 3172 8747 0.3626386189550703 EMBEZZLEMENT 837 2984 0.2804959785522788 SUICIDE 309 1291 0.2393493415956623 FRAUD 9291 41348 0.22470252491051562 NON-CRIMINAL 52553 236928 0.2218100013506213 ROBBERY 11385 54467 0.20902564855784236 EXTORTION 152 729 0.2085048010973937 BAD CHECKS 180 924 0.19480519480519481 ARSON 701 3875 0.1809032258064516 BURGLARY 14414 91067 0.1582790692567011 VANDALISM 12940 114718 0.11279834027789885 SUSPICIOUS OCC 8573 79087 0.10839961055546424 LARCENY/THEFT 40048 477975 0.08378680893352163 VEHICLE THEFT 10248 126228 0.08118642456507273 RECOVERED VEHICLE 575 8688 0.06618324125230203

In [0]:
resolve = spark.sql("""
with resolve as (select Category, resolution, count(*) as tally
from sf_crime group by 1,2)
select a.Category, a.resolution, a.tally, b.total 
from resolve a 
left join (select * from crime_total) b
on a.Category = b.Category 
order by b.total desc, a.tally desc
""")
display(resolve)

Category resolution tally total LARCENY/THEFT NONE 437927 477975 LARCENY/THEFT ARREST, BOOKED 25136 477975 LARCENY/THEFT ARREST, CITED 10463 477975 LARCENY/THEFT NOT PROSECUTED 1600 477975 LARCENY/THEFT UNFOUNDED 1221 477975 LARCENY/THEFT COMPLAINANT REFUSES TO PROSECUTE 471 477975 LARCENY/THEFT DISTRICT ATTORNEY REFUSES TO PROSECUTE 467 477975 LARCENY/THEFT EXCEPTIONAL CLEARANCE 442 477975 LARCENY/THEFT PROSECUTED BY OUTSIDE AGENCY 156 477975 LARCENY/THEFT PSYCHOPATHIC CASE 47 477975 LARCENY/THEFT LOCATED 44 477975 LARCENY/THEFT PROSECUTED FOR LESSER OFFENSE 1 477975 OTHER OFFENSES ARREST, BOOKED 115427 301874 OTHER OFFENSES ARREST, CITED 94817 301874 OTHER OFFENSES NONE 86815 301874 OTHER OFFENSES UNFOUNDED 1715 301874 OTHER OFFENSES PROSECUTED BY OUTSIDE AGENCY 718 301874 OTHER OFFENSES DISTRICT ATTORNEY REFUSES TO PROSECUTE 635 301874 OTHER OFFENSES COMPLAINANT REFUSES TO PROSECUTE 604 301874 OTHER OFFENSES NOT PROSECUTED 433 301874 OTHER OFFENSES EXCEPTIONAL CLEARANCE 361 301874 OTHER OFFENSES PSYCHOPATHIC CASE 212 301874 OTHER OFFENSES LOCATED 119 301874 OTHER OFFENSES PROSECUTED FOR LESSER OFFENSE 18 301874 NON-CRIMINAL NONE 184375 236928 NON-CRIMINAL PSYCHOPATHIC CASE 27027 236928 NON-CRIMINAL ARREST, BOOKED 9918 236928 NON-CRIMINAL UNFOUNDED 7398 236928 NON-CRIMINAL ARREST, CITED 3220 236928 NON-CRIMINAL COMPLAINANT REFUSES TO PROSECUTE 1432 236928 NON-CRIMINAL EXCEPTIONAL CLEARANCE 945 236928 NON-CRIMINAL LOCATED 866 236928 NON-CRIMINAL NOT PROSECUTED 796 236928 NON-CRIMINAL PROSECUTED BY OUTSIDE AGENCY 476 236928 NON-CRIMINAL DISTRICT ATTORNEY REFUSES TO PROSECUTE 470 236928 NON-CRIMINAL PROSECUTED FOR LESSER OFFENSE 5 236928 ASSAULT NONE 104543 167042 ASSAULT ARREST, BOOKED 49246 167042 ASSAULT ARREST, CITED 7027 167042 ASSAULT DISTRICT ATTORNEY REFUSES TO PROSECUTE 2051 167042 ASSAULT COMPLAINANT REFUSES TO PROSECUTE 1797 167042 ASSAULT UNFOUNDED 1000 167042 ASSAULT EXCEPTIONAL CLEARANCE 632 167042 ASSAULT PSYCHOPATHIC CASE 403 167042 ASSAULT NOT PROSECUTED 136 167042 ASSAULT PROSECUTED BY OUTSIDE AGENCY 102 167042 ASSAULT LOCATED 93 167042 ASSAULT PROSECUTED FOR LESSER OFFENSE 12 167042 VEHICLE THEFT NONE 115980 126228 VEHICLE THEFT ARREST, BOOKED 5064 126228 VEHICLE THEFT UNFOUNDED 4789 126228 VEHICLE THEFT ARREST, CITED 185 126228 VEHICLE THEFT DISTRICT ATTORNEY REFUSES TO PROSECUTE 70 126228 VEHICLE THEFT COMPLAINANT REFUSES TO PROSECUTE 53 126228 VEHICLE THEFT EXCEPTIONAL CLEARANCE 35 126228 VEHICLE THEFT LOCATED 20 126228 VEHICLE THEFT NOT PROSECUTED 19 126228 VEHICLE THEFT PROSECUTED BY OUTSIDE AGENCY 11 126228 VEHICLE THEFT PROSECUTED FOR LESSER OFFENSE 1 126228 VEHICLE THEFT PSYCHOPATHIC CASE 1 126228 DRUG/NARCOTIC ARREST, BOOKED 97926 117821 DRUG/NARCOTIC NONE 10256 117821 DRUG/NARCOTIC ARREST, CITED 9046 117821 DRUG/NARCOTIC UNFOUNDED 219 117821 DRUG/NARCOTIC NOT PROSECUTED 139 117821 DRUG/NARCOTIC PSYCHOPATHIC CASE 51 117821 DRUG/NARCOTIC COMPLAINANT REFUSES TO PROSECUTE 49 117821 DRUG/NARCOTIC EXCEPTIONAL CLEARANCE 37 117821 DRUG/NARCOTIC PROSECUTED BY OUTSIDE AGENCY 35 117821 DRUG/NARCOTIC LOCATED 24 117821 DRUG/NARCOTIC DISTRICT ATTORNEY REFUSES TO PROSECUTE 24 117821 DRUG/NARCOTIC PROSECUTED FOR LESSER OFFENSE 15 117821 VANDALISM NONE 101778 114718 VANDALISM ARREST, BOOKED 8734 114718 VANDALISM ARREST, CITED 2967 114718 VANDALISM UNFOUNDED 434 114718 VANDALISM COMPLAINANT REFUSES TO PROSECUTE 264 114718 VANDALISM PSYCHOPATHIC CASE 196 114718 VANDALISM DISTRICT ATTORNEY REFUSES TO PROSECUTE 191 114718 VANDALISM EXCEPTIONAL CLEARANCE 83 114718 VANDALISM NOT PROSECUTED 34 114718 VANDALISM LOCATED 25 114718 VANDALISM PROSECUTED BY OUTSIDE AGENCY 12 114718 WARRANTS ARREST, BOOKED 93092 99821 WARRANTS NONE 5482 99821 WARRANTS ARREST, CITED 885 99821 WARRANTS UNFOUNDED 132 99821 WARRANTS PROSECUTED BY OUTSIDE AGENCY 43 99821 WARRANTS DISTRICT ATTORNEY REFUSES TO PROSECUTE 36 99821 WARRANTS PSYCHOPATHIC CASE 36 99821 WARRANTS LOCATED 34 99821 WARRANTS EXCEPTIONAL CLEARANCE 30 99821 WARRANTS COMPLAINANT REF

#####Analysis

From the bar chart above showing the percentage of resolution rate, we can see that prostitution, warrants and driving under the influence has the highest rates of resolution among all 37 categories, with resolution rates above 94%. 

However, from the other side of the bar chart, we can see that recovered vehicle, vehicle theft, larceny/theft has the lowest rats of resolution below 10%, followed by suspicious ooc, vandalism, with resoluation rates both below 15%. 

Why are so many of these types of crimes remain unresolved? Is it due to a lack of police capability or legal restrictions on these types? I believe this question should be answered by both parties. 

From the side of police department, I suggest them investigate into those unresolved cases, to see if there's a problem of laissez-faire attitute or ability. If there's a downplay on those cases, for example, taking it for granted that theft is hard to traced and thus making no effort to resolve it, then the police should rectify the situation. If there's a need for applying advanced techniques or skills, the department in charge should fix it. 

As for the policy makers, I suggest them investigate on the motivation behind those crime. If there's a difficulty for the police to resolve those cases (for either reasons), then it would be best to resolve it before it happens by adjusting relevant policies.

Notice that larceny/theft is the most committed crime, vandalism at 7th place, vehicle theft at 8th. There should be placed more emphasis on these types of crime with least resolution rate, either to enhance the police's technique/skills for solving such crimes or enforce more restrictions/punishments on those crimes.

The second pie chart demonstrates the resolution records of top 4 most committed crimes, so as to provide more directly visualization comparing crime number and resolution rate. 

As can be seen, the resolution rate is shockingly low for these most frequent crimes with more than 60% of them unresolved in three of the four most frequent crimes. So there's really an urgent need to improve this situation.

### Conclusion

####1. Motivation

When considering about whether to reside in a certain place or not, one of the most important factors is to research on the safety and crime rate of the area. Most people would prefer a safer place to live, especially for a mid- or long-term residence, or someone plan to run their own business. 

For every new UC Berkeley student coming from other area, he/she must have heard about the safety concerns around the campus: drug, homeless people, etc. As an international student from even another country, I remember the feeling of anxiety and uncertainty over the potential safety issue. Even though I have been here for over 1 year and completed my study safe and sound, I thought it would still be helpful to analyze the crime data, in an attempt  to understand what's going on around this place and how could residents/tourists/shop owners stay cautious and how could policies be further enhanced to improve the police force. 

I choose SF data because it is a world-famous metropolis with a population over 880K, and would also hopefully be the next place I were about to live.

####2. Methods

1) Data processing

  To begin with, in order to perform analysis on SF crime data from 2003-2018 of a large size more than 551MB, I utilized **Spark SQL** to load and manipulate the dataset. 

  The reason is that Spark offers a parallel method to process the data and thus leads to a much faster running speed than using Python pandas or Hadoop. **Spark SQL can directly read from multiple sources and ensures fast execution of existing Hive queries**. Moveover, **SQL can provide a easy way to load data without checking the many APIs from PySpark**, so it is a good way to get start. **After using Spark SQL to load data, one can use PySpark DataFrame format to get and transform the data into some representations that can be used to train machine learning model in the future.**
On deciding to use Spark SQL instead of PySpark DataFrame method, I compared the performance of both methods and noticed that SQL was faster.

2) Data visualization

  After data cleaning and processing, I utilized the built-in visualization tool in Datacricks by display() function for basic visualization need, as well as **Tableau** for more advanced demonstration purpose. I selected the features of interest such as categories, time, districts of crimes to visualize so that I could better understand and analyze the potential properties and relations among different features. 
  
  I also chose the most suitable form of plot from various types in order to fulfill the analysis need. For example, I used a **line chart and bar chart of month vs crime number and grouped lines by different year, so as to display clearly the variation of crime rate within one year and also across different years**. I also used **pie charts of percentage of resolution in each category of crime and sorted by crime frequency, showing clearly the resolution condition of the most frequent crime**. 
  
  When displaying the crime event w.r.t category and time (hour) in top 3 dangerous district, I plotted a **stacked bar chart combining with line chart and divided by districts, so that it is easy to compare hourly difference of number and type of crimes as well as that across different districts**.
  
How to select the best way for visualization makes a huge difference and is of great importance in conveying the insights from data analysis results.

####3. Analysis and insights
1. From the bar chart and line chart of count of crime in 2015-2018, I noticed a decreasing crime in 2018. From the charts of 15 years, I found an increasing trend of crime since 2011, but the first four months of 2018 so far has the lowest number of crime over the years.

2. From the bar chart of hour vs count of crime of different categories, and line plot of hour vs count of crime in a certain day, I found that most of crimes occur during lunch or in the afternoon in 14-22pm, and early morning has the least amount of crime, especially in 3-7am. So it is suggested to travel in the morning and assign more police in the evening.

3. From the bar chart and pie chart of categories vs resolution rate, Prostitution, Warrants and Driving under influence are most resolved, while Recovered vehicle, Vehicle theft and Larceny/theft are least resolved. Moveover, some categories with the lowest resolution rate are also the most frequently committed crimes, which should be fixed urgently and paid more effort and attention to by the police and policy makers.